src/internal/connector/exchange/container_resolver.go

```diff
@@ -40,50 +48,147 @@ type containersEnumerator interface {
 	ctx context.Context,
 	folderID string,
 ) (*path.Builder, *path.Builder, error) {
-	return cr.idToPath(ctx, folderID, 0)
+	ctx = clues.Add(ctx, "container_id", folderID)
+
+	c, ok := cr.cache[folderID]
+	if !ok {
+		return nil, nil, clues.New("folder not cached").WithClues(ctx)
+	}
+
+	p := c.Path()
+	if p == nil {
+		return nil, nil, clues.New("folder has no path").WithClues(ctx)
+	}
+
+	return p, c.Location(), nil
+}
+
+func (cr *containerResolver) refreshContainer(
+	ctx context.Context,
+	id string,
+) (graph.CachedContainer, bool, error) {
+	ctx = clues.Add(ctx, "refresh_container_id", id)
+	logger.Ctx(ctx).Debug("refreshing container")
+
+	if cr.refresher == nil {
+		return nil, false, clues.New("nil refresher")
+	}
+
+	c, err := cr.refresher.refreshContainer(ctx, id)
+	if err != nil && graph.IsErrDeletedInFlight(err) {
+		logger.Ctx(ctx).Debug("container deleted")
+		return nil, true, nil
+	} else if err != nil {
+		// This is some other error, just return it.
+		return nil, false, clues.Wrap(err, "refreshing container").WithClues(ctx)
+	}
+
+	return c, false, nil
 }
 
 func (cr *containerResolver) idToPath(
 	ctx context.Context,
 	folderID string,
 	depth int,
-) (*path.Builder, *path.Builder, error) {
+) (*path.Builder, *path.Builder, bool, bool, error) {
 	ctx = clues.Add(ctx, "container_id", folderID)
 
 	if depth >= maxIterations {
-		return nil, nil, clues.New("path contains cycle or is too tall").WithClues(ctx)
+		return nil, nil, false, false, clues.New("path contains cycle or is too tall").WithClues(ctx)
 	}
 
 	c, ok := cr.cache[folderID]
 	if !ok {
-		return nil, nil, clues.New("folder not cached").WithClues(ctx)
+		c, shouldDelete, err := cr.refreshContainer(ctx, folderID)
+		if err != nil {
+			return nil, nil, false, false, clues.Wrap(err, "fetching uncached container")
+		}
+
+		if shouldDelete {
+			logger.Ctx(ctx).Debug("fetching uncached folder showed it was deleted")
+			return nil, nil, false, shouldDelete, err
+		}
+
+		if err := cr.addFolder(c); err != nil {
+			return nil, nil, false, false, clues.Wrap(err, "adding new folder").WithClues(ctx)
+		}
+
+		// Retry populating this container's paths.
+		// TODO(ashmrtn): May want to bump the depth here just so we don't get stuck
+		// retrying too much if for some reason things keep moving around?
+		pth, loc, _, shouldDelete, err := cr.idToPath(ctx, folderID, depth)
+		if err != nil {
+			err = clues.Wrap(err, "retry populating uncached folder")
+		}
+
+		return pth, loc, false, shouldDelete, err
 	}
 
 	p := c.Path()
 	if p != nil {
-		return p, c.Location(), nil
+		return p, c.Location(), true, false, nil
 	}
 
-	parentPath, parentLoc, err := cr.idToPath(
+	parentPath, parentLoc, parentCached, shouldDelete, err := cr.idToPath(
 		ctx,
 		ptr.Val(c.GetParentFolderId()),
 		depth+1)
 	if err != nil {
-		return nil, nil, clues.Wrap(err, "retrieving parent folder")
+		return nil, nil, true, false, clues.Wrap(err, "retrieving parent folder")
```

Comment:
I'm a bit confused by the first bool return val.  on line 148 it's called `parentCached`, implying that the parent was found in the cache (or, maybe, added to the cache?).  But it only gets evaluated to `true` if: an error occurred in idToPath for the parent container, an error occurred refreshing the current container if _its_ parent wasn't cached, or we should delete the current container whether or not its parent was cached.  Or, of course, if we found the parent path, and it was cached, and we shouldn't delete the current container.  _That_ outcome makes sense.  But the prior conditions... not so much.

I'm not sure if the true returns in error/delete cases are unintentional flags that end up benign in the current implementation, or if the value is being intentionally overloaded, contrary to its naming, to achieve a certain outcome.  Either case makes me uneasy, because it feels like something that'll be easy to mix up later on.

Reply:
so there's a broader question of how consistent we want things to be that this PR doesn't really address. Consider the case where we have `f1/f2/f3` and we've added `f2` and `f3` to the cache. When we go to populate the paths we'll find that `f1` is missing and try to go fetch it. There's a few different outcomes possible given the desired flow we have right now:
* f1 was deleted, f2 was moved
* f1 is there, f2 was moved
* f1 is there, f2 wasn't moved

For all of those cases, fetching only f1 and f2 is enough to fixup things. However, if we fetch `f2` there's now the question of if we should fetch any of it's children and how will we know we need to fetch every child of `f2` (we have no guaranteed order of iteration when populating paths). It's possible the first time we encounter a child of `f2` will be the time we fail for `f1` and thus we'll refresh `f1`, `f2`, and `f3`. However, if we also have `f1/f2/f4` then we won't refresh `f4` because that path is populated in a different iteration and `f1` will be cached for that iteration

`parentCached` is supposed to help determine if the folder we failed on was cached or not. If it wasn't then we fetch the immediate child of that folder but no others. If it was cached then we don't do anything extra